[LLVM] Add `LLVM_<proj>_RUNTIME_TARGETS` to set targets per-project

jhuber6 opened the pull request:
Summary:
Currently, the runtimes build allow users to optionally build for
multiple target architectures via the `LLVM_RUNTIME_TARGETS` option.
Once problem is that this applies to every single runtime the user has
enabled, when it's possible that we may wish to enable a different set
for just one target.

The motivating example here is for handling GPU targets as
cross-compiling. We want to be able to perform something like
`LLVM_RUNTIME_TARGETS=amdgcn-amd-amdhsa;nvptx64-nvidia-cuda` to build
runtimes targeting the GPU. However, the current support would force the
other runtimes to be built for the GPU, when this definitely will not
work.

This patch attempts to work around this in a generic way by letting
individual targets overload the set of enabled runtimes triples. The
expected use would be to enable something like the following for
targeting the GPU `libc` and in the future the GPU `libcxx`.

```
-DLLVM_ENABLE_RUNTIMES='openmp;libcxx;libcxx-abi;libc'
-DLLVM_LIBC_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'
```

petrhosek commented:
You can already do the following:
```
-DLLVM_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'
-DRUNTIMES_nvptx64-nvidia-cuda_LLVM_ENABLE_RUNTIMES='libc'
-DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES='libc'
```
This is already [used in existing builds](https://github.com/llvm/llvm-project/blob/bc66e0cf9feb8f3c38fa2e33850b6c6e5477fb21/clang/cmake/caches/Fuchsia-stage2.cmake#L136). Is that sufficient for your use case?

jhuber6 commented:
> You can already do the following:
> 
> ```
> -DLLVM_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'
> -DRUNTIMES_nvptx64-nvidia-cuda_LLVM_ENABLE_RUNTIMES='libc'
> -DRUNTIMES_amdgcn-amd-amdhsa_LLVM_ENABLE_RUNTIMES='libc'
> ```
> 
> This is already [used in existing builds](https://github.com/llvm/llvm-project/blob/bc66e0cf9feb8f3c38fa2e33850b6c6e5477fb21/clang/cmake/caches/Fuchsia-stage2.cmake#L333). Is that sufficient for your use case?

Will that allow the other targets to still be built? This approach seems more straightforward honestly, but if it's sufficient I could make it work.

jhuber6 commented:
So, I tired it out and it doesn't seem to have the semantics that I need. I want to be able to specify that only a specific project is built with the desired architecture, that means omitting it from the standard targets if present. That doesn't happen with the above `-DRUNTIMES_nvptx64-nvidia-cuda_LLVM_ENABLE_RUNTIMES='libc'` method, because it  will then cause other runtimes like `openmp` or `libunwind` to be built for the GPU which is not desired. Additionally, that syntax is a lot more difficult to read and not as simple to put in a default build.

petrhosek commented:
I'm not I follow, why are other runtimes like `openmp` and `libunwind` being buit for the GPU target if they're not listed in `RUNTIMES_nvptx64-nvidia-cuda_LLVM_ENABLE_RUNTIMES='libc'`? Can you post the full set of options you're using?

The reason I'm not a fan of the proposed solution is that it seems to be specifically tailored to libc and I'm worried that it won't generalize. Specifically, runtimes are typically layered on top of each other, that is `libunwind` is built on top of `libc`, `libcxxabi` is built on top of `libunwind`, `libcxx` is built on top of `libcxxabi`, `compiler-rt` is built on top of `libcxx`. Whereas `LLVM_LIBC_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'` makes sense since `libc` is the bottommost layer, something like `LLVM_LIBCXXABI_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'` likely won't ever be of any use.

In your description, you also mentioned `LLVM_ENABLE_PROJECTS='openmp;libcxx;libcxx-abi;libc'` but that's not even a supported configuration, in particular `libcxx` and `libcxxabi` only support `LLVM_ENABLE_RUNTIMES`, putting them in `LLVM_ENABLE_PROJECTS` will result in an error (that will soon be the case for all runtimes). Can you explain what the intended goal there is? I'd like to see if there's another way we can support this.

jhuber6 commented:
> I'm not sure I follow, why are other runtimes like `openmp` and `libunwind` even being built if they're not listed in `RUNTIMES_nvptx64-nvidia-cuda_LLVM_ENABLE_RUNTIMES='libc'`? Can you post the full set of options you're using?
>

What I meant is that it required passing `LLVM_RUNTIMES_TARGETS` which would then apply it to all my builds, when I want them to all be default. The option was just like this to test.

```
-DLLVM_ENABLE_RUNTIMES='libc;libcxx;libunwind;openmp'
-DLLVM_RUNTIME_TARGETS='default;x86_64-pc-linux-gnu'
-DRUNTIMES-x86_64-pc-linux-gnu_LLVM_ENABLE_RUNTIMES=libc
```

And the issue was that it was still building `default` for `libc`.
 
> The reason I'm not a fan of the proposed solution is that it seems to be specifically tailored to libc and I'm worried that it won't generalize. Specifically, runtimes are typically layered on top of each other, that is `libunwind` is built on top of `libc`, `libcxxabi` is built on top of `libunwind`, `libcxx` is built on top of `libcxxabi`, `compiler-rt` is built on top of `libcxx`. Whereas `LLVM_LIBC_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'` makes sense since `libc` is the bottommost layer, something like `LLVM_LIBCXXABI_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'` likely won't ever be of any use.
>

That's easily doable if the user manually configures the override to include either `default` or the other values in `LLVM_ENABLE_RUNTIMES`. I pretty much just need a way to enable additional targets on a per-project basis. It needs to be able to prevent the default build as well. If I were doing this and wanted the others to be built on top of it, I would just do
```
-DLLVM_RUNTIME_TARGETS=default
-DLLVM_LIBC_RUNTIME_TARGETS=default;amdgcn-amd-amdhsa
```

I think the semantics there are straightforward enogh.

> In your description, you also mentioned `LLVM_ENABLE_PROJECTS='openmp;libcxx;libcxx-abi;libc'` but that's not even a supported configuration, in particular `libcxx` and `libcxxabi` only support `LLVM_ENABLE_RUNTIMES`, putting them in `LLVM_ENABLE_PROJECTS` will result in an error (that will soon be the case for all runtimes). Can you explain what the intended goal there is? I'd like to see if there's another way we can support this.

That's a typo, I meant `LLVM_ENABLE_RUNTIMES` but my brain typed the other thing.

jhuber6 commented:
Also, like you said, this might not be explicitly useful beyond `libc` for now. I could simply make it a specific `LLVM_LIBC_RUNTIME_TARGETS` instead? But I'm planning ahead because I think in the future we will want a libc++ for the GPU that builds off of the libc for the GPU.

jhuber6 commented:
Would it be possible to make this move along faster if I just implemented it as `LLVM_LIBC_RUNTIME_TARGETS` and handled it specially there? I have the GPU `libc` rewrite done so I'm hoping to land it and get to work updating everything else since that patch solves pretty much every remaining issue I had before I could say the GPU `libc` was "done".

jhuber6 commented:
Also @petrhosek, I've noticed that the non-default targets don't use `PASSTHROUGH_PREFIXES`. Is there a reason for this? It's important to be able to pass arguments to the runtime. IMHO we should *always* pass `${canon_name}_` prefixes. I also have a need for `CUDA` so that users can pick up the CUDA Toolkit directory.

JonChesterfield reviewed (dismissed):
I'm unconvinced by the runtime layer argument. Specifically, I'm interested in building libcxx for amdgpu without building libc for it, and it's plausible that libcxx on GPUs will not involve libunwind at all.

Likewise building compiler-rt for x64 and not nvptx, while building libc for nvptx and not x64 is probably a default configuration.

I would very much prefer cmake let us specify specific combinations instead of allowing a few blessed stacks.

GPU targets as cross compiling things, instead of as magic weird things with their own configuration, is the right direction.

JonChesterfield commented:
@petrhosek please shout out if you want to block this in the current form - I have no doubt that the cmake could be better, but that's always true, and this would really help the gpu project. When someone sees a better way to express the same capability we'll be delighted to adopt it.

petrhosek commented on llvm/runtimes/CMakeLists.txt:
```diff
@@ -269,11 +270,12 @@ endfunction()
   set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${LLVM_BINARY_DIR}/runtimes/${name}/Components.cmake)
 
   set(runtime_names ${RUNTIME_NAMES})
+  set(LLVM_ENABLE_RUNTIMES ${ARG_RUNTIMES})
```
It's confusing to set this variable here since it may imply that it is somehow related to the loop below while it isn't, this should be moved to line 341.

jhuber6 replied:
Good point, done.

petrhosek commented on llvm/runtimes/CMakeLists.txt:
```diff
@@ -236,6 +236,7 @@ function(runtime_default_target)
   endif()
 
   set_enable_per_target_runtime_dir()
+  set(LLVM_ENABLE_RUNTIMES ${ARG_RUNTIMES})
```
Previously, we would passthrough `LLVM_ENABLE_RUNTIMES` as a global variable, but now we always pass in `RUNTIMES` as a function argument, and then set `LLVM_ENABLE_RUNTIMES` as a local variable shadowing the global one and rely on the passthrough mechanism.

That's quite obscure and potentially error prone, I think we should stop using the passthrough machinery and instead pass `LLVM_ENABLE_RUNTIMES` as an extra argument.

jhuber6 replied:
I think I tired adding it to `CMAKE_ARGS` but it was being overridden by the passthrough version like you said. Changing `PASSTHROUGH` was a much bigger chance so I tried to keep it small. Maybe I could set `LLVM_ENABLE_RUNTIMES` to `""` to make it more obvious.

petrhosek commented:
@jhuber6 @JonChesterfield Can you give me a more concrete example of how you intend to build the full set of runtimes (including compiler-rt, libc and libcxx) for GPU using this mechanism? Maybe that would help me better understand the motivation.

The reason I'm asking is because I'm still unconvinced by the simplicity arguments. The reason for pushing this feature is to simplify the runtimes build compared to current implementation that requires multiple variables, but in my experience, building runtimes typically requires a number of variables regardless, so saving one variable for setting the list of runtimes makes it difficult to justify more complexity in the runtimes build for me.

For reference, here's an example from our runtimes build for Linux (which includes compiler-rt, libunwind, libcxxabi and libcxx):

https://github.com/llvm/llvm-project/blob/c098f2d1669fb1ec6a5eac81277bd16b334ec956/clang/cmake/caches/Fuchsia-stage2.cmake#L146-L191

jhuber6 commented:
> @jhuber6 @JonChesterfield Can you give me a more concrete example of how you intend to build the full set of runtimes (including compiler-rt, libc and libcxx) for GPU using this mechanism? Maybe that would help me better understand the motivation.
> 
> The reason I'm asking is because I'm still unconvinced by the simplicity arguments. The reason for pushing this feature is to simplify the runtimes build compared to current implementation that requires multiple variables, but in my experience, building runtimes typically requires a number of variables regardless, so saving one variable for setting the list of runtimes makes it difficult to justify more complexity in the runtimes build for me.
> 
> For reference, here's an example from our runtimes build for Linux (which includes compiler-rt, libunwind, libcxxabi and libcxx):
> 
> https://github.com/llvm/llvm-project/blob/c098f2d1669fb1ec6a5eac81277bd16b334ec956/clang/cmake/caches/Fuchsia-stage2.cmake#L146-L191

The issue with the current support is that it does not provide a way to disable a project for one of the runtimes targets. Right now I will say that a majority of users do not want the LLVM CPU `libc`, but let's say they want the GPU `libc` and the CPU `libc++`.

Right now the suggested method would be.
```
-DLLVM_ENABLE_RUNTIMES="libc;libc++;openmp"
-DLLVM_RUNTIMES_TARGETS="default;amdgcn-amd-amdhsa"
-DRUNTIMES_amdgcn-amd-amdhsa_ENABLE_RUNTIMES='libc'
```
This would have the desired behavior of making the `libc` project compile under the AMDGPU triple. But it would still make it present under the `default` target, which is not the desired behavior. The syntax is also more difficult to understand, but we could possible just make another user facing variable that just sets that stuff automatically. So, what I wanted was the ability to "per-project" remove some of the other projects. So the above would be
```
-DLLVM_ENABLE_RUNTIMES="libc;libc++;openmp"
-DLLVM_LIBC_RUNTIMES_TARGETS="amdgcn-amd-amdhsa"
```
If there were some alternate way to specify "I want this project to only build with X runtimes" then I would also take that.

petrhosek commented:
> Right now the suggested method would be.
> 
> ```
> -DLLVM_ENABLE_RUNTIMES="libc;libc++;openmp"
> -DLLVM_RUNTIMES_TARGETS="default;amdgcn-amd-amdhsa"
> -DRUNTIMES_amdgcn-amd-amdhsa_ENABLE_RUNTIMES='libc'
> ```
> 
> This would have the desired behavior of making the `libc` project compile under the AMDGPU triple. But it would still make it present under the `default` target, which is not the desired behavior.

If you don't want `libc` being present in the `default` target, why not simply omit it from `LLVM_ENABLE_RUNTIMES`?

```
-DLLVM_ENABLE_RUNTIMES="libc++;openmp"
-DLLVM_RUNTIMES_TARGETS="default;amdgcn-amd-amdhsa"
-DRUNTIMES_amdgcn-amd-amdhsa_ENABLE_RUNTIMES="libc"
```

jhuber6 commented:
> > Right now the suggested method would be.
> > ```
> > -DLLVM_ENABLE_RUNTIMES="libc;libc++;openmp"
> > -DLLVM_RUNTIMES_TARGETS="default;amdgcn-amd-amdhsa"
> > -DRUNTIMES_amdgcn-amd-amdhsa_ENABLE_RUNTIMES='libc'
> > ```
> > This would have the desired behavior of making the `libc` project compile under the AMDGPU triple. But it would still make it present under the `default` target, which is not the desired behavior.
> 
> If you don't want `libc` being present in the `default` target, why not simply omit it from `LLVM_ENABLE_RUNTIMES`?
> 
> ```
> -DLLVM_ENABLE_RUNTIMES="libc++;openmp"
> -DLLVM_RUNTIMES_TARGETS="default;amdgcn-amd-amdhsa"
> -DRUNTIMES_amdgcn-amd-amdhsa_ENABLE_RUNTIMES="libc"
> ```

This is a little broken right now with the current logic, since we use things like 
```if ("libc" IN_LIST LLVM_ENALBE_RUNTIMES")```
in various places. That could possible be fixed if we check through all the runtime targets then.

jhuber6 commented:
> > > Right now the suggested method would be.
> > > ```
> > > -DLLVM_ENABLE_RUNTIMES="libc;libc++;openmp"
> > > -DLLVM_RUNTIMES_TARGETS="default;amdgcn-amd-amdhsa"
> > > -DRUNTIMES_amdgcn-amd-amdhsa_ENABLE_RUNTIMES='libc'
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > This would have the desired behavior of making the `libc` project compile under the AMDGPU triple. But it would still make it present under the `default` target, which is not the desired behavior.
> > 
> > 
> > If you don't want `libc` being present in the `default` target, why not simply omit it from `LLVM_ENABLE_RUNTIMES`?
> > ```
> > -DLLVM_ENABLE_RUNTIMES="libc++;openmp"
> > -DLLVM_RUNTIMES_TARGETS="default;amdgcn-amd-amdhsa"
> > -DRUNTIMES_amdgcn-amd-amdhsa_ENABLE_RUNTIMES="libc"
> > ```
> 
> This is a little broken right now with the current logic, since we use things like `if ("libc" IN_LIST LLVM_ENALBE_RUNTIMES")` in various places. That could possible be fixed if we check through all the runtime targets then.

Also I somewhat like all the used targets being present somewhere up front. Maybe we could just add a
```
-DRUNTIMES_default_DISABLE_RUNTIMES="libc"
```
or something.

jhuber6 commented:
I tried your suggested format after fixing the known places where CMake checks `if ("libc" IN_LISTS LLVM_ENABLE_RUNTIMES)` and it doesn't seem to work, I don't see the GPU triples in the runtime directory.

jhuber6 committed:
[LLVM] Add `LLVM_<proj>_RUNTIME_TARGETS` to set targets per-project

Summary:
Currently, the runtimes build allow users to optionally build for
multiple target architectures via the `LLVM_RUNTIME_TARGETS` option.
Once problem is that this applies to every single runtime the user has
enabled, when it's possible that we may wish to enable a different set
for just one target.

The motivating example here is for handling GPU targets as
cross-compiling. We want to be able to perform something like
`LLVM_RUNTIME_TARGETS=amdgcn-amd-amdhsa;nvptx64-nvidia-cuda` to build
runtimes targeting the GPU. However, the current support would force the
other runtimes to be built for the GPU, when this definitely will not
work.

This patch attempts to work around this in a generic way by letting
individual targets overload the set of enabled runtimes triples. The
expected use would be to enable something like the following for
targeting the GPU `libc` and in the future the GPU `libcxx`.

```
-DLLVM_ENABLE_PROJECTS='openmp;libcxx;libcxx-abi;libc'
-DLLVM_LIBC_RUNTIME_TARGETS='nvptx64-nvidia-cuda;amdgcn-amd-amdhsa'
```

jhuber6 commented:
Today I have learned about the joys of `;` versus `|` in CMake.

@petrhosek I somewhat understand the difficulties here. It seems that the existing support tries to do `triples=projects` while this does `projects=triples`. I think both can coexist, and have somewhat different use-cases. Do you have any more qualms about this patch? I tried getting the alternative format you provided to work but it didn't seem to do what I wanted even after patching up `libc`.

jhuber6 commented:
Any remaining concerns @petrhosek? I'd like to land this but would like to wait for your blessing.

JonChesterfield commented:
This is a prereq for pull/81921 which is a huge decrease in overall complexity of our build scripts. Hopefully @petrhosek has some more ideas for cleaning up this area that we can apply post-commit.

petrhosek commented:
I looked at all `"libc" IN_LISTS LLVM_ENABLE_RUNTIMES` cases but neither of those seemed like blockers and could be worked around by setting `LIBC_HDRGEN_ONLY=ON` and `LLVM_ENABLE_PROJECTS=libc` which IMHO is the correct solution rather than complicating the existing logic by iterating over all variables to check for `"libc" IN_LISTS RUNTIMES_<target>_LLVM_ENABLE_RUNTIMES` for every `<target>` in `LLVM_RUNTIME_TARGETS`.

I'm still not especially enthused about this feature, it adds additional complexity to the runtimes build and from what I can tell so far, you can achieve the same result with the existing implementation, and although that might require setting more options, I'm not sure it warrants the additional complexity. If there some specific things you believe aren't possible, it'd be great to provide an example so I can see if we could support those with the existing features.

If we decide to go forward with change regardless, I'd like you to setup a builder that exercises it, ideally both presubmit and postsubmit. The problem we already have is that runtimes build has too many ways to spell the same thing which is making it frustratingly difficult to make changes. For example, in case of libc I have to verify that `LLVM_ENABLE_PROJECTS=libc`, `LLVM_ENABLE_RUNTIMES=libc` and `RUNTIMES_<target>_LLVM_ENABLE_RUNTIMES=libc` still works; with this change there's yet another variation and I don't have resources to try every one of them every time I make some change to the build and I'd like to avoid a situation where I make a change and then hear back some time later that I broke a downstream user.

I'd also be curious to hear from @smeenai and @llvm-beanz as other reviewers of the runtimes build changes if they have other thoughts.

JonChesterfield commented:
Have you looked at pull/81921? That is a huge decrease in spurious complexity that we can't have without something functionally equivalent to this PR. I don't mind what horrendous cmake invocation is required to build things but I do want 81921, it lets me get rid of some really fragile infra for working with libc.

petrhosek commented:
AFAICT pull/81921 is not dependent on this change, this change just adds syntax sugar on top of the existing functionality provided by the runtimes build and so shouldn't be considered a blocker.

JonChesterfield commented:
Both `runtime=targets` and `target=runtimes` syntaxes look identically good/bad to me. I have to copy&paste cmake invocations from our documentation to get llvm to build anyway.

Keeping the current syntax and shipping the better behaviour today is much better than spending another week trying to make the syntax look different while the behavioural change bitrots.

jhuber6 commented:
So, I think my main problem with the other support is that it forces explicitly adding `default` to the list of enabled. As far as I'm aware, the vast majority of users want the default target when building. Maybe we could just have an alternate way to se them that implies `default` so users don't need to change as much if they want to copy paste something.

smeenai commented:
Apologies for the radio silence here, I'm about to go on an extended leave soon and have had lots to wrap up before then. I generally agree with Petr that adding another configuration option (and in particular one that goes from project -> target, rather than target -> project) is adding extra testing burden that we'd ideally find a way to avoid. I haven't been able to look at the other diffs to weigh that against the simplifications permitted by this though.
